packages/core/src/browser/shell/shell-layout-restorer.ts

```diff
@@ -147,30 +209,68 @@ export class ShellLayoutRestorer implements CommandContribution {
     /**
      * Creates the layout data from its string representation.
      */
-    protected inflate(layoutData: string): Promise<ApplicationShell.LayoutData> {
-        const pending: Promise<Widget | undefined>[] = [];
-        const result = JSON.parse(layoutData, (property: string, value) => {
+    protected async inflate(layoutData: string): Promise<ApplicationShell.LayoutData> {
+        const parseContext = new ShellLayoutRestorer.ParseContext();
+        const layout = this.parse<ApplicationShell.LayoutData>(layoutData, parseContext);
+
+        // tslint:disable-next-line:no-any
+        let layoutVersion: number | any;
+        try {
+            layoutVersion = 'version' in layout && Number(layout.version);
+        } catch { /* no-op */ }
+        if (typeof layoutVersion !== 'number' || Number.isNaN(layoutVersion)) {
+            throw new Error('could not resolve a layout version');
+        }
+        if (layoutVersion !== applicationShellLayoutVersion) {
+            if (layoutVersion < applicationShellLayoutVersion) {
+                console.warn(`Layout version ${layoutVersion} is behind current layout version ${applicationShellLayoutVersion}, trying to migrate...`);
+            } else {
+                console.warn(`Layout version ${layoutVersion} is ahead current layout version ${applicationShellLayoutVersion}, trying to load anyway...`);
+            }
+            console.info(`Please use '${resetLayoutCommand.label}' command if the layout looks bogus.`);
```

Comment:
This is not really visible to users, isn't it?

Do we know if a migration failed? If so, wdyt of notifying with the option to trigger the layout reset?

Reply:
Do you mean it should be `console.log`? 

I can remove it, i've added just for a case if layout does not look ok, but migrations did not fail or layout restoration, for example if migrations are actually missing, and a user had a look at console and get a hint how to recover.

Reply:
> Do we know if a migration failed?

that's what I mean.

it would be a bad UX if something happen there and we cannot notify the user (via notification.) the issue we've seen when testing this PR, shows that a migration may break the layout somehow. do we really expect users to open JS console to inspect the logs? if possible I'd rather show a message and offer an option to resolve the situation. if that's out of scope here, I'd propose to tackle this separately.

Reply:
If a migration failed then layout will be completely reset. Do you want me to show a user notification as well?

If migrations are not failed, but layout is bogus, then it is a programming error.

Reply:
> If migrations are not failed, but layout is bogus, ...

that we cannot detect? is that correct?

Reply:
> that we cannot detect? is that correct?

yes, i mean that we just missed something, like i've missed `problems` widget. It was small thing this time, another time a widget can be completely unusable. We can detect it only by testing and fix by adding migrations.